### PR TITLE
Azure waf

### PR DESCRIPTION
**Pull request by JasBeilin**
## Status
- [x] In Progress

## Related Issues
fixes: https://github.com/demisto/etc/issues/30463

## Description
Adding a new Azure Web Application Firewall integration

## Minimum version of Demisto
- [x] 5.0.0

## Does it break backward compatibility?
   - [x] No

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+        base_url = f'https://management.azure.com/subscriptions/{subscription_id}/'
+        self.subscription_id = subscription_id
+        self.resource_group_name = resource_group_name
+        client_args = client_args = {
+            'self_deployed': True,  # We always set the self_deployed key as True because when not using a self
+                                    # deployed machine, the DEVICE_CODE flow should behave somewhat like a self deployed
+                                    # flow and most of the same arguments should be set, as we're !not! using OProxy.
+            'auth_id': app_id,
+            'token_retrieval_url': 'https://login.microsoftonline.com/organizations/oauth2/v2.0/token',
+            'grant_type': DEVICE_CODE,  # disable-secrets-detection
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:44]: still needed?
```suggestion
            'grant_type': DEVICE_CODE,
```

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+    @logger
+    def http_request(self, method: str, url_suffix: str = None, full_url: str = None, params: Dict = None,
+                     data: Dict = None, resp_type: str = 'json', return_empty_response: bool = False):
+        if not params:
+            params = {}
+        if not full_url:
+            params['api-version'] = API_VERSION
+
+        return self.ms_client.http_request(method=method,  # disable-secrets-detection
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:63]:
```suggestion
        return self.ms_client.http_request(method=method,
```

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+            'ok_codes': (200, 201, 202, 204)
+        }
+
+        self.ms_client = MicrosoftClient(**client_args)
+
+    @logger
+    def http_request(self, method: str, url_suffix: str = None, full_url: str = None, params: Dict = None,
+                     data: Dict = None, resp_type: str = 'json', return_empty_response: bool = False):
+        if not params:
+            params = {}
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:59]: maybe we can se it in the func args?

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+        )
+
+    def get_policy_list_by_subscription_id(self) -> Dict:
+        return self.http_request(
+            method='GET',
+            url_suffix='/providers/Microsoft.Network/ApplicationGatewayWebApplicationFirewallPolicies'
+        )
+
+    def update_policy_upsert(self, policy_name: str, resource_group_name: str, data: Dict) -> Dict:
+
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:93]:
```suggestion
```

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+        return self.http_request(
+            method='PUT',
+            url_suffix=f'/resourceGroups/{resource_group_name}/providers/'
+                       f'Microsoft.Network/ApplicationGatewayWebApplicationFirewallPolicies/{policy_name}',
+            data=data
+        )
+
+    def delete_policy(self, policy_name: str, resource_group_name: str):
+
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:102]:
```suggestion
```

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+''' MAIN FUNCTION '''
+
+
+def main() -> None:
+    """main function, parses params and runs command functions
+
+    :return:
+    :rtype:
+    """
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:302]:
```suggestion
    """main function, parses params and runs command functions
    """
```

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+    )
+
+    demisto.debug(f'Command being called is {demisto.command()}')
+    try:
+
+        if command == 'test-module':
+            raise ValueError("Please run `!azure-waf-auth-start` and `!azure-waf-auth-complete` to log in."
+                             " For more details press the (?) button.")
+        if command == 'azure-waf-test':
+            # return_results(test_connection(client, **params))
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:331]: ?

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    """
+    demisto_commands = {'azure-waf-policy-get': policy_get_command,
+                        'azure-waf-policy-list-all-in-subscription': policy_get_list_by_subscription_command,
+                        'azure-waf-policy-upsert': policy_upsert_command,
+                        'azure-waf-policy-delete': policy_delete_command,
+                        'azure-waf-auth-start': start_auth,
+                        'azure-waf-auth-complete': complete_auth,
+                        'azure-waf-test': test_connection,
+
+                        }
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:311]: please indent this.

**Review by bakatzir — COMMENTED** [review]
please also add docstring to the funcs(no need to add to the ones in the client)

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py (hunk 3 above)
Comment (JasBeilin) [reply Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:59]: It is bad practice as far as I know to put mutable object in a function's arguments, though in this specific case it does not matter, I prefer to keep it as None.

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.yml [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+      name: verbose
+      predefined:
+      - 'true'
+      - ' false'
+      required: false
+      secret: false
+    - default: false
+      defaultValue: '10'
+      description: Number of policies to be shown. (This will only affect visualised
+        data, not context).
```
[the remark this answers is not in the document]
Comment (JasBeilin) [reply Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.yml]: The limit now enforces the context as well as the markdown.

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+    Otherwise, retrieves all policies within the resource group.
+    """
+
+    policy_name: str = args.get('policy_name', '')
+    resource_group_name: str = args.get('resource_group_name', client.resource_group_name)
+    verbose = True if args.get("verbose", "false") == "true" else False
+    limit = str(args.get("limit"))
+    if not limit.isdigit():
+        raise Exception("please provide a numeric limit")
+    limit = int(limit)
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:147]: why would it happen? and let's add a default. e.g:
```suggestion
    limit = int(args.get("limit", '20'))
```

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+                         raw_response=policies)
+    return res
+
+
+def policy_get_list_by_subscription_command(client: AzureWAFClient, **args: Dict[str, Any]) -> CommandResults:
+    """
+    Retrieve all policies within the subscription id.
+    """
+    policies: List[Dict] = []
+    verbose = True if args.get("verbose", "false") == "true" else False
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:174]:
```suggestion
    verbose = args.get("verbose", "false") == "true"
```

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+    limit = str(args.get("limit"))
+    if not limit.isdigit():
+        raise Exception("please provide a numeric limit")
+    limit = int(limit)
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:178]: same

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+
+
+def policy_delete_command(client: AzureWAFClient, **args: Dict[str, str]):
+    """
+    Gets a policy name and resource group (or taking instance's default)
+    and delete the policy from the resource group.
+    """
+    policy_name = str(args.get('policy_name', ''))
+    resource_group_name = str(args.get('resource_group_name', client.resource_group_name))
+    policy_id = f'/subscriptions/{client.subscription_id}/resourceGroups/{resource_group_name}/' \
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:246]: let's check that together.

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+
+    client = AzureWAFClient(
+        app_id=params.get('app_id', ''),
+        subscription_id=params.get('subscription_id', ''),
+        resource_group_name=params.get('resource_group_name', ''),
+        verify=not params.get('insecure', False),
+        proxy=params.get('proxy', False),
+    )
+
+    demisto.debug(f'Command being called is {demisto.command()}')
```
Comment (bakatzir) [inline Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:355]:
```suggestion
    demisto.debug(f'Command being called in Azure WAF is: {command}')
```

**Review by bakatzir — COMMENTED** [review]
Nice work. Relevant for all cmds:
1. let's revise the `.isdigit()` and limit default values
2. having an empty arr, e.g `policies` and extending it is great. but how will the HR look with no policies?
consider doing something like: https://github.com/demisto/content/blob/master/Packs/Securonix/Integrations/Securonix/Securonix.py#L814

File: Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py (hunk 13 above)
Comment (bakatzir) [reply Packs/AzureWAF/Integrations/AzureWAF/AzureWAF.py:246]: so let's add comment that this is a unique id.

**Comment by ShirleyDenkberg** [comment]
@bakatzir Doc review is completed.

File: Packs/AzureWAF/pack_metadata.json [hunk 15]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Azure WAF",
+    "description": "TO DO",
```
Comment (bakatzir) [inline Packs/AzureWAF/pack_metadata.json:3]: please do :)

**Review by bakatzir — APPROVED** [review]

**Pull request merged**